### ISSUE TITLE
Canonical, split-invariant CFG post-order

### ISSUE THREAD

**Issue by stoklund** [bytecodealliance/cranelift#145]
In e0fd5252d5c9d0fb0eaae5f33a2d55b197c33ad4 we landed the function `DominatorTree::recompute_split_ebb()` which updates the dominator tree and its cached CFG post-order after an EBB split. It looks like the way we repair the post order is not correct. Consider this example:

```
ebb0:
    brnz v1, ebb1
    brnz v2, ebb2
    brnz v3, ebb3
    brnz v4, ebb4
    jump ebb5
```

If we assume that all of ebb0's successors are not visited before ebb0 in the DFS, we get the following post order: ebb5, ebb4, ebb3, ebb2, ebb1, ebb0.

Now assume we split the EBB by inserting the ebb99 header:

```
ebb0:
    brnz v1, ebb1
    brnz v2, ebb2
    jump ebb99
ebb99
    brnz v3, ebb3
    brnz v4, ebb4
    jump ebb5
```

The post order is now: ebb5, ebb4, ebb3, **ebb99**, ebb2, ebb1, ebb0. Two claims:

1. The current implementation of `recompute_split_ebb()` is not correct because it inserts the new EBB adjacent to ebb0. As the example shows, that is not always the right insertion spot.
2. The post order currently computed by the dominator tree is *split invariant*. By this I mean that the effect on the post order of splitting an EBB is just to insert the new EBB header somewhere. The relative order of existing EBBs does not change. It would be good to prove this property more strictly.

Some implementation notes for fixing this:

- Assuming the post order really is split invariant, the strategy in `recompute_split_ebb()` of inserting the new EBB somewhere in the existing post order is correct. It is only the insertion position that needs to be fixed.
- The new EBB will always appear before the old EBB header in the post order.
- The successors to the old EBB (ebb0 in the example) do not necessarily come before the old EBB in the post order. It is possible for any of the to be visited before ebb0 along another CFG path, which would cause them to appear *after* ebb0 in the post order.
- If one of the successors after the split appear before ebb0 in the post order, the correct insertion point is after the first such successor.

**Comment by KiChjang** [comment]
So I'm currently taking a look at this issue, and I have 2 questions:

1. Under what circumstances would you call `Layout::split_ebb`? I can see that cursor.rs calls this function, but under what scenario would you insert an EBB in between instructions?
2. Is it guaranteed that the instruction right before the split is a terminator? I understand that EBBs are explicitly defined to have such property, but since I don't know when EBBs are being split, I cannot say for sure.

**Comment by KiChjang** [comment]
An addenum to the 2nd question: if it is currently not guaranteed that the instruction before the split is a terminator, would it be beneficial to add a check that ensures the instruction must be a terminator, before the insertion of EBB can be made?

**Comment by stoklund** [comment]
Two situations where an EBB might be split:

1. The loop analysis pass may split an EBB to make sure that every EBB is part of at most two natural loops, see #106. This is the simplest possible split where we just insert a jump from the old EBB to the new EBB.
2. The legalizer may need to split an EBB and insert one or more new EBBs. See for example the functions `expand_select()`, `expand_cond_trap()`, and Intel's `expand_minmax()`.

We don't maintain a valid dominator tree during legalization at the moment, so you can focus on the simple split done by loop analysis.

Ad 2) The verifier will enforce the invariant that every EBB ends with a terminator instruction. When modifying the IR, it may temporarily be possible to have an EBB without a terminator or a terminator instruction in the middle of an EBB. However, by the time functions like `recompute_split_ebb()` and `cfg.recompute_ebb()` are called, these invariants can be assumed to hold again.

> An addenum to the 2nd question: if it is currently not guaranteed that the instruction before the split is a terminator, would it be beneficial to add a check that ensures the instruction must be a terminator, before the insertion of EBB can be made?

I don't think so. This kind of check will have the effect of forcing you to do the IR modifications in a certain order, and that can make the code more convoluted. If you want the IR to be valid at all times, you would need to provide a terminator instruction to the EBB splitting function since you can't legally have a terminator instruction in the middle of the EBB either.

To summarize: The IR can be temporarily invalid while it is being modified, but by the time `recompute_split_ebb()` is called, it is guaranteed to be valid again. You can add assertions to check that if you like.

**Comment by stoklund** [comment]
For general reading, I recommend [Notes on graph algorithms used in optimizing compilers](http://www.cs.umb.edu/~offner/files/flow_graph.pdf) and [A simple, fast dominance algorithm](http://www.cs.rice.edu/~keith/EMBED/dom.pdf).

**Comment by KiChjang** [comment]
I'd like to check my understanding for the postorder. Given the following:

```
ebb0:
  v1 = iconst.i64 0
  jump ebb1

ebb1:
  brz v1, ebb0
  jump ebb2

ebb2:
  brz v1, ebb0
  brz v1, ebb0
  brz v1, ebb0
  jump ebb100

ebb100:
  return
```

Would the above program have a postorder of ebb100, ebb2, ebb0, ebb1?

**Comment by stoklund** [comment]
No, ebb100, ebb2, ebb1, ebb0. I'd recommend looking at chapter 1 of the notes I linked above.
